Implementation for hook to define phases of playback during git commands

Wiseqube opened the pull request.

Wiseqube commented:
This is running as expected so far, we'll of course have to have the correct script to make use of this functionality!

viktordick commented on perfact/zodbsync/subcommand.py:
```diff
@@ -150,12 +151,80 @@ def gitexec(func):
         - play back changed objects (diff between old and new HEAD)
         - unstash
         """
+        def _playback_paths(self, paths):
```
Why do we need this embedded function? Could we pull this out of `def gitexec`?

Wiseqube replied:
Shouldn't be a problem, I think this was due to the iterative development. Basically started in the `wrapper` but soon moved it out but kept it contained inside of gitexec

viktordick commented on perfact/zodbsync/subcommand.py:
```diff
@@ -150,12 +151,80 @@ def gitexec(func):
+                    dryrun=dryrun,
+                )
+
+                if not dryrun and phase_cmd and os.path.isfile(phase_cmd):
```
Please turn this around and by using a negated condition and `continue`.

viktordick reviewed (changes requested):
Mostly LGTM. Two suggestions for simplifying the code, see below.

Please also adjust the example `config.py` that is part of the repo as well as the `README.md`.

Wiseqube commented:
@viktordick I've adjusted the files according to your review, I also added an entry in the readme, the config and the changelog! Additionally, which is mainly due to the fact that I saw a deprecation warning during the running of the checks I've updated the major versions of the github actions used in our workflows!

viktordick commented on perfact/zodbsync/subcommand.py:
```diff
@@ -139,6 +140,75 @@ def check_repo(self):
+        if playback_hook and os.path.isfile(playback_hook):
+            proc = subprocess.Popen(
+                playback_hook, stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
```
Didn't see that earlier - why do you capture `stderr`? I think it would be better if the caller would see any error messages on their command line.

viktordick replied:
For now, I will merge this since I want to build on top of this. I think we should continue the discussion about what to to with `stderr` here, but that can be done in a later PR. For now, the hook script must not write anything to `stderr` in the case of success.

viktordick commented on perfact/zodbsync/subcommand.py:
```diff
@@ -223,6 +285,17 @@ def wrapper(self, *args, **kwargs):
                         self.logger.exception("Unable to show diff")
 
                 self.abort()
+                # if we are not in dryrun we can't be sure we havent already
```
Wait, what is doing this here? Can you explain?

viktordick reviewed (approved):
LGTM
